src/Std/Data/DTreeMap/Basic.lean

```diff
@@ -0,0 +1,355 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Markus Himmel, Paul Reichert
+-/
+prelude
+import Std.Data.DTreeMap.Internal.WF.Def
+
+/-!
+# Dependent tree maps
+
+This file develops the type `Std.DTreeMap` of dependent tree maps.
+
+Lemmas about the operations on `Std.DTreeMap` will be available in the
+module `Std.Data.DTreeMap.Lemmas`.
+
+See the module `Std.Data.DTreeMap.Raw` for a variant of this type which is safe to use in
+nested inductive types.
+-/
+
+set_option autoImplicit false
+set_option linter.missingDocs true
+
+universe u v
+
+variable {α : Type u} {β : α → Type v} {cmp : α → α → Ordering}
+private local instance : Coe (Type v) (α → Type v) where coe γ := fun _ => γ
+
+namespace Std
+
+/--
+Dependent tree maps.
+
+A tree map stores an assignment of keys to values. It depends on a comparator function that
+defines an ordering on the keys and provides efficient order-dependent queries, such as retrieval
+of the minimum or maximum.
+
+To ensure that the operations behave as expected, the comparator function `cmp` should satisfy
+certain laws that ensure a consistent ordering:
+
+* If `a` is less than (or equal) to `b`, then `b` is greater than (or equal) to `a`
+and vice versa (see the `OrientedCmp` typeclass).
+* If `a` is less than or equal to `b` and `b` is, in turn, less than or equal to `c`, then `a`
+is less than or equal to `c` (see the `TransCmp` typeclass).
+
+Keys for which `cmp a b = Ordering.eq` are considered the same, i.e., there can be only one entry
+with key either `a` or `b` in a tree map. Looking up either `a` or `b` always yields the same entry,
+if any is present. The `get` operations of the _dependent_ tree map additionally require a
+`LawfulEqCmp` instance to ensure that `cmp a b = .eq` always implies `a = b`, so that their
+respective value types are equal.
+
+To avoid expensive copies, users should make sure that the tree map is used linearly.
+
+Internally, the tree maps are represented as size-bounded trees, a type of self-balancing binary
+search tree with efficient order statistic lookups.
+
+These tree maps contain a bundled well-formedness invariant, which means that they cannot
+be used in nested inductive types. For these use cases, `Std.DTreeMap.Raw` and
+`Std.DTreeMap.Raw.WF` unbundle the invariant from the tree map. When in doubt, prefer
+`DTreeMap` over `DTreeMap.Raw`.
+-/
+structure DTreeMap (α : Type u) (β : α → Type v) (cmp : α → α → Ordering := by exact compare) where
+  /-- Internal implementation detail of the tree map. -/
+  inner : DTreeMap.Internal.Impl α β
+  /-- Internal implementation detail of the tree map. -/
+  wf : letI : Ord α := ⟨cmp⟩; inner.WF
+
+namespace DTreeMap
+open Internal (Impl)
+
+/--
+Creates a new empty tree map. It is also possible to
+use the empty collection notations `∅` and `{}` to create an empty tree map.
```

Comment:
I think here we have a difference between hash maps and tree maps. In the hash map case, there is a legitimate reason to use `empty`, namely to initialize a hash map with a given initial capacity. For the tree map, there is no such thing, so I think we should recommend to always use `∅`, have a simp lemma that rewrites `empty` to `∅` and state all lemmas only for `∅`.

Reply:
I rewrote the docstring as follows:

```lean
/--
Creates a new empty tree map. It is also possible and recommended to
use the empty collection notations `∅` and `{}` to create an empty tree map. `simp` replaces
`empty` with `∅`.
-/
```

Although this can also be read as a recommendation of `{}`, I tested that `simp` treats both notations the same as expected, so I believe this is fine.